crates/shared/src/trade_finding/oneinch.rs

```diff
@@ -127,8 +134,22 @@ impl Inner {
         })
     }
 
+    /// Returns the current 1Inch smart contract as the `spender`. Caches that value for 60 seconds
+    /// to avoid unnecessary requests.
     async fn spender(&self) -> Result<H160, TradeError> {
-        Ok(self.api.get_spender().await?.address)
+        let spender = *self.spender_cache.read().unwrap();
+        let is_recent =
+            |updated_at| Instant::now().duration_since(updated_at) < self.spender_max_age;
+        match spender {
```

Comment:
The way this is implemented, becausae we're not holding the lock during the update, multiple threads can observe that the cache is outdated and will all fetch the new value. This can be avoided by holding the lock throughout the update. I think this is fine because the wait time is the same. Either way a requests needs to be made but by waiting we only make one request.